File: viewer/src/model/trace.ts
// The trace file's shape, and nothing else.
//
// This module parses and indexes. It derives no state, classifies no event and computes no
// recovery arithmetic, because the trace already carries its own conclusions: every event records
// the resulting client state and headline numbers. A viewer that recomputed them would be a second
// implementation of a C++ state machine, written in TypeScript by someone reading the first, and
// when the two disagreed the picture would be wrong with nothing to say so.
//
// So the rule for this whole app: if a number is not in the file, it is not drawn.

export type Layer = "venue" | "chaos" | "wire" | "client";

export interface TraceEvent {
  readonly kind: "event";
  /** Which source packet the run had reached. The scrubber's x-axis. */
  readonly i: number;
  /** Nanoseconds on the run's manual clock. Simulated, and the ledger says so. */
  readonly t: number;
  readonly layer: Layer;
  readonly event: string;
  readonly line: number;
  /** Half-open sequence range the event is about, matching recovery::sequence_range. */
  readonly first: number;
  readonly end: number;
  readonly attempt: number;
  readonly detail: number;
  readonly reason: string;
  /** The client state *after* this event. Read, never inferred. */
  readonly state: string;
  readonly delivered_through: number;
  readonly missing: number;
  readonly holes: number;
  /**
   * The outstanding holes themselves, as `[first, end)` pairs: bounded at four by the writer.
   *
   * Present because the viewer needed to *draw* them and the alternative was accumulating them from the
   * gap events, which is reconstructing recovery state in TypeScript. The format grew instead. When
   * `holes` exceeds this array's length, say "and N more" rather than drawing a lie.
   */
  readonly gaps: ReadonlyArray<readonly [number, number]>;

  /**
   * The top of the order book the delivered messages have built, at this moment.
   *
   * The second time the format grew for the viewer, and the rule working rather than an exception to it. A viewer
   * that applied price levels itself would be a second implementation of an order book written in TypeScript by
   * somebody reading the C++, and the strongest thing this project asserts is that the book after loss and repair
   * equals the book that lost nothing, which a third implementation could only muddy.
   *
   * Prices are signed integers at four implied decimals: 208,700 is $20.8700. Zero means no book on that side.
   */
  readonly best_bid: number;
  readonly best_bid_size: number;
  readonly best_ask: number;
  readonly best_ask_size: number;
  /** How many levels exist, so a drawing can say "and N more" rather than imply the book is one level deep. */
  readonly bid_levels: number;
  readonly ask_levels: number;
  readonly traded_shares: number;
}

export interface ScheduledFault {
  readonly op: string;
  readonly first_packet: number;
  readonly packet_count: number;
  readonly parameter: number;
  readonly detail: number;
}

export type LimitStatus = "measured" | "not-measurable" | "simulated";

export interface Limit {
  readonly claim: string;
  readonly status: LimitStatus;
  readonly note: string;
}

export interface RunHeader {
  readonly kind: "run";
  readonly schema: string;
  readonly seed: number;
  readonly messages: number;
  readonly packets: number;
  readonly session: number;
  readonly mode: "recovering" | "glimpse";
  readonly staleness_messages: number;
  readonly lines: number;
  readonly schedule: readonly ScheduledFault[];
  readonly limits: readonly Limit[];
}

export interface RunSummary {
  readonly kind: "summary";
  readonly events: number;
  readonly events_dropped: number;
  readonly messages_delivered: number;
  readonly messages_delivered_twice: number;
  readonly messages_missing: number;
  readonly retransmit_requests: number;
  /**
   * How many messages those requests covered, as distinct from how many requests there were.
   *
   * The two answer different questions and the difference is interesting: a second line that fills the
   * middle of a hole splits one gap into two, so it can produce more requests while asking for fewer
   * messages. The format grew a field rather than have the viewer sum event ranges.
   */
  readonly retransmit_messages: number;
  readonly retransmits_served: number;
  readonly retransmit_refusals: number;
  readonly snapshot_requests: number;
  readonly unfillable_messages: number;

  /**
   * The book every published message builds, with nothing lost.
   *
   * What the run's own book has to equal. Carried so the page can state the invariant: "this is the book that
   * would have existed": rather than showing a bid and an ask a reader has to interpret. Computed by applying the
   * published bodies in order, not by a second recovery run: the reference is *what the venue sent*, and a
   * reference that also went through recovery would be comparing recovery to itself.
   */
  readonly reference_bid: number;
  readonly reference_ask: number;
  readonly reference_traded: number;
  readonly final_state: string;
  /** False when the recorder filled: the trace is a prefix, and the UI must say so. */
  readonly complete: boolean;
}

export interface Trace {
  readonly header: RunHeader;
  readonly events: readonly TraceEvent[];
  readonly summary: RunSummary;
  /** Highest packet index seen, so the scrubber has a range without scanning twice. */
  readonly lastIndex: number;
}

export class TraceFormatError extends Error {}

/**
 * Parses JSONL. Strict on purpose: a viewer that silently skipped a line it did not understand
 * would draw an incomplete run and look like a complete one, which is the failure this whole
 * project is about.
 */
export function parseTrace(text: string): Trace {
  const lines = text.split("\n").filter((line) => line.trim().length > 0);
  if (lines.length < 2) {
    throw new TraceFormatError("a trace needs at least a header and a summary");
  }

  let header: RunHeader | undefined;
  let summary: RunSummary | undefined;
  const events: TraceEvent[] = [];
  let lastIndex = 0;

  for (const [at, line] of lines.entries()) {
    const parsed: unknown = JSON.parse(line);
    if (typeof parsed !== "object" || parsed === null || !("kind" in parsed)) {
      throw new TraceFormatError(`line ${at + 1} has no kind`);
    }
    const record = parsed as { kind: string };
    switch (record.kind) {
      case "run":
        header = parsed as RunHeader;
        break;
      case "summary":
        summary = parsed as RunSummary;
        break;
      case "event": {
        const event = parsed as TraceEvent;
        events.push(event);
        lastIndex = Math.max(lastIndex, event.i);
        break;
      }
      default:
        throw new TraceFormatError(`line ${at + 1}: unknown kind "${record.kind}"`);
    }
  }

  if (header === undefined) {
    throw new TraceFormatError("no run header");
  }
  if (summary === undefined) {
    throw new TraceFormatError("no summary: the trace was truncated");
  }
  if (!header.schema.startsWith("dfr-trace/")) {
    throw new TraceFormatError(`unsupported schema "${header.schema}"`);
  }
  return { header, events, summary, lastIndex };
}

File: tools/trace.cpp
// Records one complete run of the library as JSONL, for a viewer to read.
//
//   trace [--seed N] [--messages N] [--faults N] [--lines 1|2] [--glimpse] [--staleness N]
//         [--out FILE]
//
// One JSON object per line: a header describing the run, one line per event, and a summary. The
// format is a deliverable, not a debug dump — a trace is a deterministic function of the seed, so
// it can be committed next to the test that produced it and diffed when behaviour changes. "Here is
// the three-fault schedule that breaks it" becomes a file somebody can open.
//
// The viewer must contain no domain logic
// --------------------------------------
// Every event carries the resulting client state and headline numbers, so drawing any moment needs
// one line and no replay. A viewer that reconstructed state from the event sequence would be a
// second implementation of the state machine, written in another language by someone reading the
// first — and when the two disagreed, the picture would be wrong with nothing to say so.
//
// The honesty ledger is in the data
// -------------------------------
// The header carries a `limits` array: which numbers this run measured and which cannot be measured
// on the hardware available at all. It is generated rather than written into a README so that it
// cannot drift from what the run actually did.

#include "support/traced_drivers.hpp"

#include <cstdio>
#include <cstring>
#include <string>

namespace {

using dfr_tools::run_mode;
using dfr_tools::run_options;

struct cli {
  run_options run{};
  const char* out{nullptr};
};

bool parse(int argc, char** argv, cli& into) {
  for (int i = 1; i < argc; ++i) {
    const auto next = [&]() -> const char* {
      return i + 1 < argc ? argv[++i] : nullptr;
    };
    if (std::strcmp(argv[i], "--seed") == 0) {
      const char* value = next();
      if (value == nullptr) {
        return false;
      }
      into.run.seed = std::strtoull(value, nullptr, 10);
    } else if (std::strcmp(argv[i], "--messages") == 0) {
      const char* value = next();
      if (value == nullptr) {
        return false;
      }
      into.run.messages = std::strtoull(value, nullptr, 10);
    } else if (std::strcmp(argv[i], "--faults") == 0) {
      const char* value = next();
      if (value == nullptr) {
        return false;
      }
      into.run.faults = static_cast<std::uint32_t>(std::strtoul(value, nullptr, 10));
    } else if (std::strcmp(argv[i], "--staleness") == 0) {
      const char* value = next();
      if (value == nullptr) {
        return false;
      }
      into.run.staleness_messages = std::strtoull(value, nullptr, 10);
    } else if (std::strcmp(argv[i], "--lines") == 0) {
      const char* value = next();
      if (value == nullptr) {
        return false;
      }
      into.run.lines = std::strtoull(value, nullptr, 10);
      if (into.run.lines == 0 || into.run.lines > 2) {
        std::fprintf(stderr, "trace: --lines must be 1 or 2\n");
        return false;
      }
    } else if (std::strcmp(argv[i], "--glimpse") == 0) {
      into.run.mode = run_mode::glimpse;
      if (into.run.staleness_messages == 0) {
        into.run.staleness_messages = 20;  // enough to land in the gap
      }
    } else if (std::strcmp(argv[i], "--out") == 0) {
      into.out = next();
      if (into.out == nullptr) {
        return false;
      }
    } else {
      std::fprintf(stderr, "trace: unrecognised argument %s\n", argv[i]);
      return false;
    }
  }
  return true;
}

// Minimal JSON writing, by hand and on purpose: a dependency for this would be a dependency in the
// build for the sake of twelve fields, and every string written here is either a fixed name or an
// enumerator name from the library.
void write_header(std::FILE* out, const cli& options,
                  const dfr_tools::run_summary& summary,
                  std::size_t published) {
  std::fprintf(out,
               "{\"kind\":\"run\",\"schema\":\"dfr-trace/1\",\"seed\":%llu,"
               "\"messages\":%zu,\"packets\":%zu,\"session\":%u,\"mode\":\"%s\","
               "\"staleness_messages\":%llu,\"lines\":%zu,\"schedule\":[",
               static_cast<unsigned long long>(options.run.seed),
               options.run.messages, published, dfr_tools::kTraceSession,
               options.run.mode == run_mode::glimpse ? "glimpse" : "recovering",
               static_cast<unsigned long long>(options.run.staleness_messages),
               options.run.lines);

  for (std::size_t i = 0; i < summary.schedule.size(); ++i) {
    const auto& fault = summary.schedule[i];
    std::fprintf(out,
                 "%s{\"op\":\"%s\",\"first_packet\":%llu,\"packet_count\":%u,"
                 "\"parameter\":%llu,\"detail\":%u}",
                 i == 0 ? "" : ",", dfr::chaos::to_string(fault.op).data(),
                 static_cast<unsigned long long>(fault.first_packet),
                 fault.packet_count,
                 static_cast<unsigned long long>(fault.parameter), fault.detail);
  }

  // The honesty ledger, generated so it cannot drift from the run.
  std::fprintf(
      out,
      "],\"limits\":["
      "{\"claim\":\"messages delivered exactly once\",\"status\":\"measured\","
      "\"note\":\"counted per sequence in this run\"},"
      "{\"claim\":\"gaps reported are exactly those that never arrived\","
      "\"status\":\"measured\",\"note\":\"checked against what the harness handed over\"},"
      "{\"claim\":\"reproducible from the seed\",\"status\":\"measured\","
      "\"note\":\"manual clock, seeded prng, no wall clock anywhere in the pipeline\"},"
      "{\"claim\":\"tick-to-trade latency\",\"status\":\"not-measurable\","
      "\"note\":\"cloud VMs only: no PMU counters, no NIC hardware timestamping\"},"
      "{\"claim\":\"IGMP snooping and querier behaviour\",\"status\":\"not-measurable\","
      "\"note\":\"AWS does not support multicast on an ordinary VPC\"},"
      "{\"claim\":\"timing numbers in this trace\",\"status\":\"simulated\","
      "\"note\":\"a manual clock advanced by the driver, not a measurement\"}"
      "]}\n");
}

void write_event(std::FILE* out, const dfr::trace::event& event) {
  std::fprintf(out,
               "{\"kind\":\"event\",\"i\":%llu,\"t\":%lld,\"layer\":\"%s\","
               "\"event\":\"%s\",\"line\":%u,\"first\":%llu,\"end\":%llu,\"attempt\":%u,"
               "\"detail\":%llu,\"reason\":\"%s\",\"state\":\"%s\","
               "\"delivered_through\":%llu,\"missing\":%llu,\"holes\":%llu}\n",
               static_cast<unsigned long long>(event.packet_index),
               static_cast<long long>(event.time_ns),
               dfr::trace::name_of(dfr::trace::layer_of(event.kind)).data(),
               dfr::trace::name_of(event.kind).data(), event.line,
               static_cast<unsigned long long>(event.first_sequence),
               static_cast<unsigned long long>(event.end_sequence), event.attempt,
               static_cast<unsigned long long>(event.detail),
               dfr::to_string(event.reason).data(),
               dfr::recovery::name_of(
                   static_cast<dfr::recovery::client_state>(event.client_state))
                   .data(),
               static_cast<unsigned long long>(event.delivered_through),
               static_cast<unsigned long long>(event.messages_missing),
               static_cast<unsigned long long>(event.outstanding_ranges));
}

void write_summary(std::FILE* out, const dfr_tools::run_summary& summary,
                   const dfr_tools::trace_recorder& trace) {
  std::fprintf(
      out,
      "{\"kind\":\"summary\",\"events\":%zu,\"events_dropped\":%llu,"
      "\"messages_delivered\":%llu,\"messages_delivered_twice\":%llu,"
      "\"messages_missing\":%llu,\"retransmit_requests\":%llu,"
      "\"retransmits_served\":%llu,\"retransmit_refusals\":%llu,"
      "\"snapshot_requests\":%llu,\"unfillable_messages\":%llu,"
      "\"final_state\":\"%s\",\"complete\":%s}\n",
      trace.size(), static_cast<unsigned long long>(trace.dropped()),
      static_cast<unsigned long long>(summary.messages_delivered),
      static_cast<unsigned long long>(summary.messages_delivered_twice),
      static_cast<unsigned long long>(summary.messages_missing),
      static_cast<unsigned long long>(summary.retransmit_requests),
      static_cast<unsigned long long>(summary.retransmits_served),
      static_cast<unsigned long long>(summary.retransmit_refusals),
      static_cast<unsigned long long>(summary.snapshot_requests),
      static_cast<unsigned long long>(summary.unfillable_messages),
      dfr::recovery::name_of(summary.final_state).data(),
      trace.dropped() == 0 ? "true" : "false");
}

}  // namespace

int main(int argc, char** argv) {
  cli options;
  if (!parse(argc, argv, options)) {
    std::fprintf(stderr,
                 "usage: trace [--seed N] [--messages N] [--faults N] "
                 "[--lines 1|2] [--glimpse] [--staleness N] [--out FILE]\n");
    return 2;
  }

  dfr_tools::trace_recorder recorder;
  std::int64_t clock_us = 0;
  const auto stream =
      dfr_tools::publish_stream(options.run.messages, recorder, clock_us);
  if (stream.empty()) {
    std::fprintf(stderr, "trace: the publisher produced nothing\n");
    return 1;
  }

  dfr_tools::run_summary summary =
      dfr_tools::run_traced(options.run, stream, recorder);

  std::FILE* out = stdout;
  if (options.out != nullptr) {
    out = std::fopen(options.out, "wb");
    if (out == nullptr) {
      std::fprintf(stderr, "trace: cannot write %s\n", options.out);
      return 1;
    }
  }

  write_header(out, options, summary, stream.size());
  for (const auto& event : recorder.events()) {
    write_event(out, event);
  }
  write_summary(out, summary, recorder);

  if (out != stdout) {
    std::fclose(out);
    std::fprintf(stderr,
                 "trace: %zu events written to %s (%llu dropped)\n",
                 recorder.size(), options.out,
                 static_cast<unsigned long long>(recorder.dropped()));
  }

  // A truncated trace is reported rather than presented as complete, for the same reason a partial
  // capture read fails inspect: a diagnostic that overstates itself is worse than none.
  return recorder.dropped() == 0 ? 0 : 1;
}
